📊 Zomato Bangalore Restaurant Trends

🎯 Objective

The aim of this hackathon is to perform data-driven analysis of restaurant trends in Bangalore using Zomato’s real-world dataset. Participants are expected to:
Clean and preprocess the data


Perform exploratory data analysis (EDA)


Merge location data for mapping


Extract actionable insights


Answer MCQ-style reasoning questions based on analysis


In [4]:
# pip installs

In [5]:
! pip install geopy

In [6]:
# Importing necessary libraries

In [7]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter
from geopy.geocoders import Nominatim
from geopy.exc import GeocoderTimedOut
from IPython.core.display import display, HTML
import time

/var/folders/5_/xq7pyg7d3bgg23t0r00sr3_80000gn/T/ipykernel_23332/3229509480.py:8: DeprecationWarning: Importing display from IPython.core.display is deprecated since IPython 7.14, please import from IPython display
  from IPython.core.display import display, HTML


In [8]:
# Opening the Dataframe

In [9]:
zomato = pd.read_csv("zomato_data.csv")

In [10]:
zomato

online_order book_table    rate  votes            rest_type  \
0              Yes        Yes   4.1/5    775        Casual Dining   
1              Yes         No   4.1/5    787        Casual Dining   
2              Yes         No   3.8/5    918  Cafe, Casual Dining   
3               No         No   3.7/5     88          Quick Bites   
4               No         No   3.8/5    166        Casual Dining   
...            ...        ...     ...    ...                  ...   
51712           No         No  3.6 /5     27                  Bar   
51713           No         No     NaN      0                  Bar   
51714           No         No     NaN      0                  Bar   
51715           No        Yes  4.3 /5    236                  Bar   
51716           No         No  3.4 /5     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                                NaN   
51713                                                NaN   
51714                                                NaN   
51715                       Cocktails, Pizza, Buttermilk   
51716                                                NaN   

                                     cuisines approx_costfor_two_people  \
0              North Indian, Mughlai, Chinese                       800   
1                 Chinese, North Indian, Thai                       800   
2                      Cafe, Mexican, Italian                       800   
3                  South Indian, North Indian                       300   
4                    North Indian, Rajasthani                       600   
...                                       ...                       ...   
51712                             Continental                     1,500   
51713                             Finger Food                       600   
51714                             Finger Food                     2,000   
51715                             Finger Food                     2,500   
51716  Finger Food, North Indian, Continental                     1,500   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [11]:
zomato.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype 
---  ------                     --------------  ----- 
 0   online_order               51717 non-null  object
 1   book_table                 51717 non-null  object
 2   rate                       43942 non-null  object
 3   votes                      51717 non-null  int64 
 4   rest_type                  51490 non-null  object
 5   dish_liked                 23639 non-null  object
 6   cuisines                   51672 non-null  object
 7   approx_costfor_two_people  51371 non-null  object
 8   listed_intype              51717 non-null  object
 9   listed_incity              51717 non-null  object
dtypes: int64(1), object(9)
memory usage: 3.9+ MB


In [12]:
# Geo Data Viz

Geographical Coordinates

In [14]:
geo = pd.read_csv("Geographical Coordinates.csv")

In [15]:
geo

listed_incity   Latitude  Longitude
0            Banashankari  12.939333  77.553982
1       Bannerghatta Road  12.952660  77.605048
2            Basavanagudi  12.941726  77.575502
3               Bellandur  12.925352  77.675941
4            Brigade Road  12.967358  77.606435
5             Brookefield  12.963814  77.722437
6                     BTM  12.916360  77.604733
7           Church Street  12.974914  77.605247
8         Electronic City  12.848760  77.648253
9             Frazer Town  12.998683  77.615525
10                    HSR  12.900563  77.649475
11            Indiranagar  12.996298  77.545278
12              Jayanagar  12.939904  77.582638
13               JP Nagar  12.909694  77.586607
14           Kalyan Nagar  13.022142  77.640337
15           Kammanahalli  13.014854  77.638203
16  Koramangala 4th Block  12.932778  77.629405
17  Koramangala 5th Block  12.934843  77.618977
18  Koramangala 6th Block  12.939025  77.623848
19  Koramangala 7th Block  12.936485  77.613478
20           Lavelle Road  12.968340  77.596507
21           Malleshwaram  13.002735  77.570325
22           Marathahalli  12.955257  77.698416
23                MG Road  12.975526  77.606790
24           New BEL Road  13.039186  77.564284
25       Old Airport Road  12.960632  77.642500

In [16]:
# Checking for null values

In [17]:
zomato.isnull().sum()

online_order                     0
book_table                       0
rate                          7775
votes                            0
rest_type                      227
dish_liked                   28078
cuisines                        45
approx_costfor_two_people      346
listed_intype                    0
listed_incity                    0
dtype: int64

In [18]:
# Data cleaning & Pre processing

In [19]:
# Regex - Removing unnecessary values

In [20]:
# Rate

In [21]:
# Method 1: Using str.replace() and str.strip()
zomato["rate"] = zomato["rate"].str.replace('/5', '').str.strip()

# Method 2: More robust version with error handling
# zomato["rate"] = zomato["rate"].apply(lambda x: str(x).replace('/5', '').strip() if pd.notna(x) else np.nan)

In [22]:
zomato

online_order book_table rate  votes            rest_type  \
0              Yes        Yes  4.1    775        Casual Dining   
1              Yes         No  4.1    787        Casual Dining   
2              Yes         No  3.8    918  Cafe, Casual Dining   
3               No         No  3.7     88          Quick Bites   
4               No         No  3.8    166        Casual Dining   
...            ...        ...  ...    ...                  ...   
51712           No         No  3.6     27                  Bar   
51713           No         No  NaN      0                  Bar   
51714           No         No  NaN      0                  Bar   
51715           No        Yes  4.3    236                  Bar   
51716           No         No  3.4     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                                NaN   
51713                                                NaN   
51714                                                NaN   
51715                       Cocktails, Pizza, Buttermilk   
51716                                                NaN   

                                     cuisines approx_costfor_two_people  \
0              North Indian, Mughlai, Chinese                       800   
1                 Chinese, North Indian, Thai                       800   
2                      Cafe, Mexican, Italian                       800   
3                  South Indian, North Indian                       300   
4                    North Indian, Rajasthani                       600   
...                                       ...                       ...   
51712                             Continental                     1,500   
51713                             Finger Food                       600   
51714                             Finger Food                     2,000   
51715                             Finger Food                     2,500   
51716  Finger Food, North Indian, Continental                     1,500   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [23]:
# Convert rate column to numeric
zomato['rate'] = pd.to_numeric(zomato['rate'],errors='coerce' )

In [24]:
# Convert approx_costfor_two_people column to numeric
zomato['approx_costfor_two_people'] = pd.to_numeric(zomato['approx_costfor_two_people'],errors='coerce')

In [25]:
# Replacing null values with Median

In [26]:
# Rate

In [27]:
zomato['rate'] = zomato['rate'].fillna(zomato['rate'].median()) 

In [28]:
# Cost

In [29]:
zomato['approx_costfor_two_people'] = zomato['approx_costfor_two_people'].fillna(zomato['approx_costfor_two_people'].median())

In [30]:
# Cleaning commas from column

In [31]:
# Convert cost column to numeric
zomato['approx_costfor_two_people'] = (
    zomato['approx_costfor_two_people']
    .astype(str)  # Convert all values to strings first
    .str.replace(',', '')  # Remove commas from numbers
    .replace({'nan': pd.NA, 'NEW': pd.NA})  # Handle special cases
    .apply(pd.to_numeric, errors='coerce')  # Final conversion to numeric
)

In [32]:
zomato['approx_costfor_two_people'] = zomato['approx_costfor_two_people'].astype('int64')

In [33]:
zomato.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  object 
 1   book_table                 51717 non-null  object 
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51490 non-null  object 
 5   dish_liked                 23639 non-null  object 
 6   cuisines                   51672 non-null  object 
 7   approx_costfor_two_people  51717 non-null  int64  
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(1), int64(2), object(7)
memory usage: 3.9+ MB


In [34]:
zomato

online_order book_table  rate  votes            rest_type  \
0              Yes        Yes   4.1    775        Casual Dining   
1              Yes         No   4.1    787        Casual Dining   
2              Yes         No   3.8    918  Cafe, Casual Dining   
3               No         No   3.7     88          Quick Bites   
4               No         No   3.8    166        Casual Dining   
...            ...        ...   ...    ...                  ...   
51712           No         No   3.6     27                  Bar   
51713           No         No   3.7      0                  Bar   
51714           No         No   3.7      0                  Bar   
51715           No        Yes   4.3    236                  Bar   
51716           No         No   3.4     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                                NaN   
51713                                                NaN   
51714                                                NaN   
51715                       Cocktails, Pizza, Buttermilk   
51716                                                NaN   

                                     cuisines  approx_costfor_two_people  \
0              North Indian, Mughlai, Chinese                        800   
1                 Chinese, North Indian, Thai                        800   
2                      Cafe, Mexican, Italian                        800   
3                  South Indian, North Indian                        300   
4                    North Indian, Rajasthani                        600   
...                                       ...                        ...   
51712                             Continental                        400   
51713                             Finger Food                        600   
51714                             Finger Food                        400   
51715                             Finger Food                        400   
51716  Finger Food, North Indian, Continental                        400   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [35]:
zomato["approx_costfor_two_people"].value_counts()

approx_costfor_two_people
400    13825
300     7576
500     4980
200     4857
600     3714
250     2959
800     2285
150     2066
700     1948
350     1763
450     1417
100      993
650      776
550      761
750      758
900      700
850      166
950       62
120       24
180       20
80        10
230       10
130        8
50         8
40         8
330        4
199        4
70         3
190        2
240        2
360        2
140        2
160        1
469        1
60         1
560        1
Name: count, dtype: int64

In [36]:
# Cleaning categorical columns

In [37]:
# Clean categorical columns
zomato['dish_liked'] = zomato['dish_liked'].fillna("Not Available")
zomato['cuisines'] = zomato['cuisines'].fillna("Other")
zomato['rest_type'] = zomato['rest_type'].fillna("Unknown")

# Verify the changes
print("Missing values after cleaning:")
print(pd.DataFrame({
    'dish_liked': zomato['dish_liked'].isna().sum(),
    'cuisines': zomato['cuisines'].isna().sum(),
    'rest_type': zomato['rest_type'].isna().sum()
}, index=['Remaining NaN Count']))

print("\nSample values:")
print(zomato[['dish_liked', 'cuisines', 'rest_type']].head())

Missing values after cleaning:
                     dish_liked  cuisines  rest_type
Remaining NaN Count           0         0          0

Sample values:
                                          dish_liked  \
0  Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1  Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2  Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                        Masala Dosa   
4                                Panipuri, Gol Gappe   

                         cuisines            rest_type  
0  North Indian, Mughlai, Chinese        Casual Dining  
1     Chinese, North Indian, Thai        Casual Dining  
2          Cafe, Mexican, Italian  Cafe, Casual Dining  
3      South Indian, North Indian          Quick Bites  
4        North Indian, Rajasthani        Casual Dining  


In [38]:
# Dropping Unnecessary columns

In [39]:
# zomato = zomato.drop(["approx_cost_for_two"], axis=1)

In [40]:
zomato

online_order book_table  rate  votes            rest_type  \
0              Yes        Yes   4.1    775        Casual Dining   
1              Yes         No   4.1    787        Casual Dining   
2              Yes         No   3.8    918  Cafe, Casual Dining   
3               No         No   3.7     88          Quick Bites   
4               No         No   3.8    166        Casual Dining   
...            ...        ...   ...    ...                  ...   
51712           No         No   3.6     27                  Bar   
51713           No         No   3.7      0                  Bar   
51714           No         No   3.7      0                  Bar   
51715           No        Yes   4.3    236                  Bar   
51716           No         No   3.4     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                      Not Available   
51713                                      Not Available   
51714                                      Not Available   
51715                       Cocktails, Pizza, Buttermilk   
51716                                      Not Available   

                                     cuisines  approx_costfor_two_people  \
0              North Indian, Mughlai, Chinese                        800   
1                 Chinese, North Indian, Thai                        800   
2                      Cafe, Mexican, Italian                        800   
3                  South Indian, North Indian                        300   
4                    North Indian, Rajasthani                        600   
...                                       ...                        ...   
51712                             Continental                        400   
51713                             Finger Food                        600   
51714                             Finger Food                        400   
51715                             Finger Food                        400   
51716  Finger Food, North Indian, Continental                        400   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [41]:
# Binary Encoding

Convert the following binary fields:
Column:  Mapping:
online_order
Yes → 1, No → 0
book_table
Yes → 1, No → 0

In [43]:
zomato['online_order'] = zomato['online_order'].map({'Yes': 1, 'No': 0})
zomato['book_table'] = zomato['book_table'].map({'Yes': 1, 'No': 0})

In [44]:
zomato

online_order  book_table  rate  votes            rest_type  \
0                 1           1   4.1    775        Casual Dining   
1                 1           0   4.1    787        Casual Dining   
2                 1           0   3.8    918  Cafe, Casual Dining   
3                 0           0   3.7     88          Quick Bites   
4                 0           0   3.8    166        Casual Dining   
...             ...         ...   ...    ...                  ...   
51712             0           0   3.6     27                  Bar   
51713             0           0   3.7      0                  Bar   
51714             0           0   3.7      0                  Bar   
51715             0           1   4.3    236                  Bar   
51716             0           0   3.4     13   Bar, Casual Dining   

                                              dish_liked  \
0      Pasta, Lunch Buffet, Masala Papad, Paneer Laja...   
1      Momos, Lunch Buffet, Chocolate Nirvana, Thai G...   
2      Churros, Cannelloni, Minestrone Soup, Hot Choc...   
3                                            Masala Dosa   
4                                    Panipuri, Gol Gappe   
...                                                  ...   
51712                                      Not Available   
51713                                      Not Available   
51714                                      Not Available   
51715                       Cocktails, Pizza, Buttermilk   
51716                                      Not Available   

                                     cuisines  approx_costfor_two_people  \
0              North Indian, Mughlai, Chinese                        800   
1                 Chinese, North Indian, Thai                        800   
2                      Cafe, Mexican, Italian                        800   
3                  South Indian, North Indian                        300   
4                    North Indian, Rajasthani                        600   
...                                       ...                        ...   
51712                             Continental                        400   
51713                             Finger Food                        600   
51714                             Finger Food                        400   
51715                             Finger Food                        400   
51716  Finger Food, North Indian, Continental                        400   

       listed_intype listed_incity  
0             Buffet  Banashankari  
1             Buffet  Banashankari  
2             Buffet  Banashankari  
3             Buffet  Banashankari  
4             Buffet  Banashankari  
...              ...           ...  
51712  Pubs and bars    Whitefield  
51713  Pubs and bars    Whitefield  
51714  Pubs and bars    Whitefield  
51715  Pubs and bars    Whitefield  
51716  Pubs and bars    Whitefield  

[51717 rows x 10 columns]

In [45]:
# Data Type Conversion

Ensure the following conversions:
rate → float


votes → integer


approx_costfor_two_people → integer

In [47]:
zomato.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51717 entries, 0 to 51716
Data columns (total 10 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   online_order               51717 non-null  int64  
 1   book_table                 51717 non-null  int64  
 2   rate                       51717 non-null  float64
 3   votes                      51717 non-null  int64  
 4   rest_type                  51717 non-null  object 
 5   dish_liked                 51717 non-null  object 
 6   cuisines                   51717 non-null  object 
 7   approx_costfor_two_people  51717 non-null  int64  
 8   listed_intype              51717 non-null  object 
 9   listed_incity              51717 non-null  object 
dtypes: float64(1), int64(4), object(5)
memory usage: 3.9+ MB


In [48]:
END

NameError: name 'END' is not defined

In [ ]:
zomato["rest_type"]

In [ ]:
# MCQ Questions

1. What is the shape of the given dataset?

51717 rows × 10 columns

2. How many restaurants serve North Indian cuisine?

In [ ]:
zomato["cuisines"].value_counts()

In [ ]:
# Count the number of restaurants that serve North Indian cuisine
# We check if 'North Indian' is in the 'cuisines' column (case-insensitive)

north_indian_count = zomato['cuisines'].dropna().apply(lambda x: 'north indian' in x.lower()).sum()
north_indian_count

3. What cuisine is most commonly offered by restaurants in Bangalore?

In [ ]:
zomato["cuisines"].value_counts()

In [ ]:
# Extract cuisines and split into individual items
cuisines = zomato['cuisines'].dropna().str.split(',\s*').explode()

# Count occurrences of each cuisine
cuisine_counts = Counter(cuisines)

# Get the most common cuisine
most_common_cuisine, count = cuisine_counts.most_common(1)[0]

# Display results
print(f"All cuisine counts:\n{cuisine_counts}")
print(f"\nMost common cuisine: '{most_common_cuisine}' (appears {count} times)")

4. Which locality in Bangalore has the highest average cost for dining (for two people)?

In [ ]:
# Recheck

In [ ]:
# Clean the 'approx_costfor_two_people' column (remove commas and convert to numeric)
zomato['approx_costfor_two_people'] = (
    zomato['approx_costfor_two_people']
    .astype(str)
    .str.replace(',', '')
    .replace('NEW', pd.NA)  # Handle 'NEW' entries as missing
    .replace('', pd.NA)     # Handle empty strings as missing
    .astype(float)
)

# Group by locality and calculate the mean cost for two people
locality_cost = (
    zomato.groupby('listed_incity')['approx_costfor_two_people']
    .mean()
    .sort_values(ascending=False)
    .dropna()
)

# Get the locality with the highest average cost
highest_cost_locality = locality_cost.idxmax()
highest_cost_value = locality_cost.max()

# Display results
print("Average cost for two people by locality:")
print(locality_cost.head(10))  # Show top 10 for context
print(f"\nLocality with the highest average cost: '{highest_cost_locality}' (₹{highest_cost_value:.2f})")

In [ ]:
# Group by locality and calculate mean cost
avg_cost_by_locality = zomato.groupby('listed_incity')['approx_costfor_two_people'].mean().sort_values(ascending=False)

# Get the locality with the highest average cost
highest_avg_cost_locality = avg_cost_by_locality.idxmax()
highest_avg_cost_value = avg_cost_by_locality.max()

print(f"The locality with the highest average cost for dining is **{highest_avg_cost_locality}** with an average of **₹{highest_avg_cost_value:.2f}** for two people.")

5. Which restaurant type has the top rating with over 1000 votes?

In [ ]:
# Recheck

In [ ]:
# # Clean the 'rate' column (extract numeric part from '4.1/5')
# zomato['rate'] = (
#     zomato['rate']
#     .astype(str)
#     .str.extract(r'(\d+\.\d+)')[0]  # Extract the numeric part (e.g., '4.1' from '4.1/5')
#     .astype(float)
# )

# # Clean the 'votes' column (ensure it's numeric)
# zomato['votes'] = pd.to_numeric(zomato['votes'], errors='coerce')

# Filter restaurants with over 1000 votes
high_vote_restaurants = zomato[zomato['votes'] > 1000]

# Group by restaurant type and calculate the mean rating
rest_type_ratings = (
    high_vote_restaurants.groupby('rest_type')['rate']
    .mean()
    .sort_values(ascending=False)
    .dropna()
)

# Get the top-rated restaurant type
top_rest_type = rest_type_ratings.idxmax()
top_rating = rest_type_ratings.max()

# Display results
print("Average ratings for restaurant types with >1000 votes:")
print(rest_type_ratings.head(10))  # Show top 10 for context
print(f"\nTop-rated restaurant type: '{top_rest_type}' (Average rating: {top_rating:.2f}/5)")

6. How much does it cost at minimum to eat out in Bangalore?

In [ ]:
# Recheck

In [ ]:
# Clean the 'approx_costfor_two_people' column:
# 1. Remove commas (e.g., '1,200' → '1200')
# 2. Convert to numeric, coercing invalid entries (like 'NEW' or empty strings) to NaN
# zomato['approx_cost_for_two'] = (
#     zomato['approx_costfor_two_people']
#     .astype(str)
#     .str.replace(',', '')
#     .replace('NEW', pd.NA)  # Treat 'NEW' as missing
#     .replace('', pd.NA)     # Treat empty strings as missing
#     .astype(float)
# )

# Find the minimum non-zero and non-NaN cost
min_cost = zomato['approx_costfor_two_people'].min()

# Display the result
print(f"The minimum cost to eat out in Bangalore (for two people) is: ₹{min_cost:.2f}")

7. What percentage of total online orders is received by restaurants in Banashankari?

In [ ]:
# Recheck

In [ ]:
# Filter restaurants in Banashankari
banashankari_data = zomato[zomato['listed_incity'] == 'Banashankari']

# Count online orders in Banashankari
banashankari_online_orders = banashankari_data['online_order'].sum()

# Count total online orders across all localities
total_online_orders = zomato['online_order'].sum()

# Calculate the percentage
percentage_banashankari = (banashankari_online_orders / total_online_orders) * 100

print(f"Banashankari receives **{percentage_banashankari:.2f}%** of the total online orders in Bangalore.")

8. Which locality has the most restaurants with over 500 votes and a rating below 3.0?

In [ ]:
# # Clean the data:
# # 1. Extract numeric part from 'rate' (e.g., '3.8/5' → 3.8)
# zomato['rate'] = (
#     zomato['rate']
#     .astype(str)
#     .str.extract(r'(\d+\.\d+)')[0]  # Extract the numeric rating
#     .astype(float)
# )

# # 2. Ensure 'votes' is numeric (handle missing/errors)
# zomato['votes'] = pd.to_numeric(zomato['votes'], errors='coerce')

# Filter restaurants with >500 votes and rating <3.0
filtered_data = zomato[
    (zomato['votes'] > 500) & 
    (zomato['rate'] < 3.0)
]

# Count restaurants by locality and find the top one
locality_counts = filtered_data['listed_incity'].value_counts()

if not locality_counts.empty:
    top_locality = locality_counts.idxmax()
    count = locality_counts.max()
    print(f"Locality with most restaurants (>500 votes & rating <3.0): '{top_locality}' ({count} restaurants)")
else:
    print("No restaurants meet the criteria (>500 votes & rating <3.0).")

# Optional: Display all localities meeting the criteria
print("\nAll localities with restaurants matching the criteria:")
print(locality_counts)

9. Which locality in Bangalore should Zomato target for expansion based on restaurant type diversity?

In [ ]:
# Clean the 'rest_type' column (split combined types like "Cafe, Casual Dining")
zomato['rest_type'] = zomato['rest_type'].str.split(',\s*')

# Explode the 'rest_type' column to count each type separately
exploded_types = zomato.explode('rest_type')

# Calculate metrics for each locality:
# 1. Number of unique restaurant types
# 2. Total number of restaurants (proxy for demand)
locality_stats = (
    exploded_types.groupby('listed_incity')
    .agg(
        unique_rest_types=('rest_type', 'nunique'),  # Count unique types
        total_restaurants=('rest_type', 'count')     # Count total entries
    )
    .sort_values(['unique_rest_types', 'total_restaurants'], ascending=False)
)

# Normalize scores (0-100) for comparison
locality_stats['diversity_score'] = (
    (locality_stats['unique_rest_types'] / locality_stats['unique_rest_types'].max()) * 50 +
    (locality_stats['total_restaurants'] / locality_stats['total_restaurants'].max()) * 50
)

# Get the top locality for expansion
top_locality = locality_stats['diversity_score'].idxmax()
top_score = locality_stats['diversity_score'].max()

# Display results
print("Top localities by restaurant type diversity and demand:")
print(locality_stats.head(10))
print(f"\nBest locality for expansion: '{top_locality}' (Diversity Score: {top_score:.1f}/100)")

10. What's the average cost difference between buffet and delivery restaurants?

In [ ]:
# Recheck

In [ ]:
# Clean the 'approx_costfor_two_people' column (remove commas, handle missing values)
zomato['approx_cost_for_two'] = (
    zomato['approx_costfor_two_people']
    .astype(str)
    .str.replace(',', '')
    .replace('NEW', pd.NA)
    .replace('', pd.NA)
    .astype(float)
)

# Filter for Buffet and Delivery restaurants
buffet_data = zomato[zomato['listed_intype'] == 'Buffet']
delivery_data = zomato[zomato['listed_intype'] == 'Delivery']

# Calculate average costs
avg_buffet_cost = buffet_data['approx_cost_for_two'].mean()
avg_delivery_cost = delivery_data['approx_cost_for_two'].mean()

# Compute the difference
cost_difference = avg_buffet_cost - avg_delivery_cost

# Display results
print(f"Average cost for Buffet (for two): ₹{avg_buffet_cost:.2f}")
print(f"Average cost for Delivery (for two): ₹{avg_delivery_cost:.2f}")
print(f"\nCost difference (Buffet - Delivery): ₹{cost_difference:.2f}")

11 . What is the maximum number of votes received by any restaurant with online ordering?

In [ ]:
# Filter restaurants with online ordering enabled
online_restaurants = zomato[zomato['online_order'] == 1]

# Find the maximum votes received
max_votes = online_restaurants['votes'].max()

print(f"The maximum number of votes received by any restaurant with online ordering is **{max_votes}**.")

12. What is the average rating of restaurants that serve both North Indian and Chinese cuisines?

In [ ]:
# # Clean the 'rate' column (extract numeric part from '4.1/5')
# zomato['rate'] = (
#     zomato['rate']
#     .astype(str)
#     .str.extract(r'(\d+\.\d+)')[0]  # Extract numeric rating
#     .astype(float)
# )

# Filter restaurants serving BOTH North Indian and Chinese cuisines
north_indian_chinese_restaurants = zomato[
    zomato['cuisines'].str.contains('North Indian', na=False) & 
    zomato['cuisines'].str.contains('Chinese', na=False)
]

# Calculate the average rating
average_rating = north_indian_chinese_restaurants['rate'].mean()

# Display results
print(f"Number of restaurants serving both cuisines: {len(north_indian_chinese_restaurants)}")
print(f"Average rating: {average_rating:.2f}/5")

13. What is the most profitable area for Zomato based on potential revenue estimation?

In [ ]:
# Recheck

In [ ]:
# Filter restaurants with online ordering (assuming online orders drive Zomato's revenue)
online_restaurants = zomato[zomato['online_order'] == 1]

# Calculate revenue potential per restaurant: (cost_for_two/2) * votes (proxy for order volume)
online_restaurants['revenue_potential'] = (online_restaurants['approx_costfor_two_people'] / 2) * online_restaurants['votes']

# Group by locality and sum revenue potential
revenue_by_locality = online_restaurants.groupby('listed_incity')['revenue_potential'].sum().sort_values(ascending=False)

# Get the most profitable locality
most_profitable_locality = revenue_by_locality.idxmax()
highest_revenue = revenue_by_locality.max()

print(f"The most profitable area for Zomato is **{most_profitable_locality}** with a revenue potential of **₹{highest_revenue:,.2f}**.")

14. If Zomato wants to reduce customer complaints, which restaurant type should they focus on?

In [ ]:
# Clean the 'rate' column - handle various formats and missing values
def clean_rating(rate):
    if pd.isna(rate):
        return None
    if isinstance(rate, str):
        if '/' in rate:
            return float(rate.split('/')[0].strip())
        elif 'NEW' in rate or '-' in rate:
            return None
    try:
        return float(rate)
    except:
        return None

zomato['cleaned_rate'] = zomato['rate'].apply(clean_rating)

# Remove rows with no valid rating
zomato = zomato.dropna(subset=['cleaned_rate'])

# Define low-rated threshold (you can adjust this)
LOW_RATING_THRESHOLD = 2.0
low_rated = zomato[zomato['cleaned_rate'] < LOW_RATING_THRESHOLD]

# If no low-rated restaurants, adjust threshold or handle differently
if len(low_rated) == 0:
    print("No restaurants found below rating threshold. Try lowering the threshold.")
    # Example: try with lower threshold
    LOW_RATING_THRESHOLD = 4.0
    low_rated = zomato[zomato['cleaned_rate'] < LOW_RATING_THRESHOLD]
    if len(low_rated) == 0:
        print("Still no restaurants found. Check your rating data.")
        exit()

# Count restaurants by type
rest_type_counts = zomato['rest_type'].value_counts().reset_index()
rest_type_counts.columns = ['rest_type', 'total_count']

low_rated_counts = low_rated['rest_type'].value_counts().reset_index()
low_rated_counts.columns = ['rest_type', 'low_rated_count']

# Merge counts
analysis = pd.merge(
    rest_type_counts,
    low_rated_counts,
    on='rest_type',
    how='left'
).fillna(0)  # Fill NA with 0 for types with no low-rated restaurants

# Calculate percentages and metrics
analysis['pct_low_rated'] = (analysis['low_rated_count'] / analysis['total_count']) * 100
analysis = analysis[analysis['total_count'] > 10]  # Only consider types with sufficient samples

# Add average ratings
avg_ratings = zomato.groupby('rest_type')['cleaned_rate'].mean().reset_index()
analysis = pd.merge(analysis, avg_ratings, on='rest_type')

# Calculate complaint risk score
analysis['complaint_risk'] = analysis['pct_low_rated'] * (5 - analysis['cleaned_rate'])
analysis = analysis.sort_values('complaint_risk', ascending=False)

# Display results
print(f"\nRestaurant types to focus on (rating threshold: {LOW_RATING_THRESHOLD}):")
print(analysis[['rest_type', 'pct_low_rated', 'cleaned_rate', 'complaint_risk']].head(10))

In [ ]:
zomato["rest_type"]

15. In which area should Zomato invest by considering high rating (rate > 4.2), high number of votes (> 500) and including online orders?

In [ ]:
# Debugging: Check value ranges
print("\nRating range:", zomato['rate'].min(), "-", zomato['rate'].max())
print("Votes range:", zomato['votes'].min(), "-", zomato['votes'].max())
print("Online order counts:\n", zomato['online_order'].value_counts())

# Adjust criteria if too strict
rating_threshold = 4.0  # Lowered from 4.2
votes_threshold = 300   # Lowered from 500

high_quality = zomato[
    (zomato['rate'] > rating_threshold) & 
    (zomato['votes'] > votes_threshold) & 
    (zomato['online_order'] == True)
]

# Debugging: Show how many restaurants meet criteria
print(f"\nRestaurants meeting criteria (rating > {rating_threshold}, votes > {votes_threshold}, online order):", len(high_quality))

if len(high_quality) > 0:
    # Group by area and calculate metrics
    investment_areas = high_quality.groupby('listed_incity').agg({
        'votes': 'sum',
        'approx_costfor_two_people': 'mean',
        'rate': 'mean',
        'rest_type': 'count'
    }).rename(columns={'rest_type': 'restaurant_count'})

    # Calculate investment score (weighted by multiple factors)
    investment_areas['investment_score'] = (
        investment_areas['votes'] * 
        investment_areas['rate'] * 
        (investment_areas['approx_costfor_two_people'] / 100)  # Normalize cost
    )

    # Sort and show results
    best_areas = investment_areas.sort_values('investment_score', ascending=False)
    
    print("\nTop areas for Zomato to invest in:")
    print(best_areas.head(10))
else:
    print("\nNo areas meet all criteria. Try relaxing the thresholds further.")
    print("Alternative approach: Looking at areas with highest average ratings...")
    
    alt_approach = zomato.groupby('listed_incity').agg({
        'rate': 'mean',
        'votes': 'mean',
        'online_order': lambda x: (x == True).mean()  # % with online ordering
    }).sort_values(['rate', 'votes'], ascending=[False, False])
    
    print(alt_approach.head(10))

In [ ]:
# 

In [ ]:
# 🗺️ Geospatial Visualization Task

In [ ]:
# Data Visualization

In [ ]:
# Restaurant Density Map

In [ ]:
# Merging Datasets

In [ ]:
merged_df = pd.merge(zomato, geo, on='listed_incity', how='left')

In [ ]:
merged_df

In [ ]:
# pip installs

In [ ]:
! pip install folium

In [ ]:
# 1. Restaurant Density Map

In [ ]:
import folium
from folium.plugins import HeatMap
import pandas as pd

# Load and prepare data (replace with your dataset)
# df = pd.read_csv('zomato_bangalore.csv')

# Clean location data - ensure we have valid latitudes/longitudes
merged_df = merged_df.dropna(subset=['Latitude', 'Longitude'])

# Create base map centered on Bangalore
bangalore_coords = [12.9716, 77.5946]
m = folium.Map(location=bangalore_coords, zoom_start=12)

# Add heatmap layer for restaurant density
heat_data = [[row['Latitude'], row['Longitude']] for index, row in merged_df.iterrows()]
HeatMap(heat_data, radius=15).add_to(m)

# Add markers for top-rated restaurants (rating > 4.0)
top_restaurants = merged_df[merged_df['rate'] > 4.0]
for idx, row in top_restaurants.iterrows():
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"{row['rest_type']} - Rating: {row['rate']}",
        icon=folium.Icon(color='green', icon='cutlery')
    ).add_to(m)

# Save and display
m.save('restaurant_density.html')
m

In [ ]:
# 2. Cuisine-Specific Distribution Maps

In [ ]:
# Create separate maps for popular cuisines
popular_cuisines = ['North Indian', 'Chinese', 'South Indian', 'Italian', 'Biryani']

for cuisine in popular_cuisines:
    # Filter restaurants for this cuisine
    cuisine_df = merged_df[merged_df['cuisines'].str.contains(cuisine, na=False)]
    
    # Create map
    cuisine_map = folium.Map(location=bangalore_coords, zoom_start=12)
    
    # Add markers
    for idx, row in cuisine_df.iterrows():
        folium.Marker(
            location=[row['Latitude'], row['Longitude']],
            popup=f"{row['rest_type']} - {cuisine}",
            icon=folium.Icon(color='blue', icon='cutlery')
        ).add_to(cuisine_map)
    
    # Add title
    title_html = f'<h3 align="center" style="font-size:16px"><b>{cuisine} Restaurants in Bangalore</b></h3>'
    cuisine_map.get_root().html.add_child(folium.Element(title_html))
    
    # Save
    cuisine_map.save(f'{cuisine.lower().replace(" ", "_")}_restaurants.html')

In [ ]:
# 3. Restaurant Type Cluster Map

In [ ]:
from folium.plugins import MarkerCluster

# Create cluster map for different restaurant types
cluster_map = folium.Map(location=bangalore_coords, zoom_start=12)

# Group by restaurant type
for rest_type in merged_df['rest_type'].value_counts().head(5).index:
    type_df = merged_df[merged_df['rest_type'] == rest_type]
    
    # Create marker cluster for this type
    marker_cluster = MarkerCluster(name=rest_type).add_to(cluster_map)
    
    for idx, row in type_df.iterrows():
        folium.Marker(
            location=[row['Latitude'], row['Longitude']],
            popup=f"{row['rest_type']} - {rest_type}",
            icon=folium.Icon(color='orange', icon='cutlery')
        ).add_to(marker_cluster)

# Add layer control
folium.LayerControl().add_to(cluster_map)
cluster_map.save('restaurant_type_clusters.html')
cluster_map

In [ ]:
# 4. Interactive Analysis Map

In [ ]:
from folium.plugins import FastMarkerCluster

# Create comprehensive analysis map
analysis_map = folium.Map(location=bangalore_coords, zoom_start=12)

# Color coding by rating
def get_color(rating):
    if rating > 4.0: return 'green'
    elif rating > 3.5: return 'blue'
    else: return 'red'

# Add all restaurants with rating-based coloring
for idx, row in merged_df.iterrows():
    folium.CircleMarker(
        location=[row['Latitude'], row['Longitude']],
        radius=5,
        popup=f"{row['rest_type']} - Rating: {row.get('rate', 'N/A')}",
        color=get_color(row.get('rate', 0)),
        fill=True,
        fill_color=get_color(row.get('rate', 0))
    ).add_to(analysis_map)

# Add heatmap overlay
HeatMap(heat_data, radius=15, name='Density Heatmap').add_to(analysis_map)

# Add layer control
folium.LayerControl().add_to(analysis_map)
analysis_map.save('restaurant_analysis.html')
analysis_map

In [ ]:
# Cuisine-Specific Map (Italian Restaurants)

In [ ]:
# 1: Filter for Italian Restaurants

In [ ]:
# Filter for Italian restaurants (case insensitive)
italian_restaurants = merged_df[merged_df['cuisines'].str.contains('Italian', case=False, na=False)]
print(f"Found {len(italian_restaurants)} Italian restaurants")

In [ ]:
# 2: Create the Base Map

In [ ]:
import folium

# Create base map centered on the city's average coordinates
avg_lat = italian_restaurants['Latitude'].mean()
avg_lon = italian_restaurants['Longitude'].mean()
m = folium.Map(location=[avg_lat, avg_lon], zoom_start=12)

In [ ]:
# 3: Add Italian Restaurant Markers

In [ ]:
# Add markers for each Italian restaurant
for idx, row in italian_restaurants.iterrows():
    # Use .get() method to safely access columns that might have different names
    name = row.get('name', row.get('restaurant_name', 'Italian Restaurant'))
    rating = row.get('rate', row.get('rating', 'N/A'))
    
    folium.Marker(
        location=[row['Latitude'], row['Longitude']],
        popup=f"{name} - Rating: {rating}",
        icon=folium.Icon(color='red', icon='pasta', prefix='fa')  # Using Font Awesome pasta icon
    ).add_to(m)

In [ ]:
# 5: Add Layer Control and Save

In [ ]:
folium.LayerControl().add_to(m)
m.save('italian_restaurants_map.html')